design/open/high-level-python-wrapper.md

```diff
@@ -0,0 +1,341 @@
+
+
+class CommonPrefix:
+    def __init__(self, repository_id: str, reference_id: str, path: str, client: Client = DefaultClient): ...
+    def exists(self) -> bool: ...
+
+
+class ObjectInfo:
+    def __init__(self, repository_id: str, reference_id: str, path: str, client: Client = DefaultClient): ...
+    def path(self) -> self: ...
+    def modified_time(self) -> datetime.datetime: ...
+    def size_bytes(self) -> int: ...
+    def content_type(self) -> Optional[str]: ...
+    def metadata(self) -> dict[str, str]: ...
+    def physical_address(self) -> str: ...
+    def delete(self): ...
+
+
+class Change(NamedTuple):
+    type: Literal(['added', 'removed', 'changed', 'conflict,', 'prefix_changed'])
+    path: str
+    path_type: Literal(['common_prefix', 'object'])
+    size_bytes: Optional[int]
+
+
+class ServerConfiguration:
+    def __init__(self, client: Client = DefaultClient): ...
+    def version(self) -> str: ...
+    def storage_config(self) -> ServerStorageConfiguration: ...
+
+
+class ServerStorageConfiguration(NamedTuple):
+    blockstore_type: str
+    pre_sign_support: bool
+    import_support: bool
+
+
+class ImportManager:
+    def __init__(self, repository_id: str, reference_id: str, client: Client = DefaultClient): ...
+    def prefix(self, object_store_uri: str, destination: str) -> ImportManager: ...
+    def object(self, object_store_uri: str, destination: str) -> ImportManager: ...
+    def start(self) -> str:
+        'start import, reporting back (and storing) a process id'
+        ...
+    def wait(self, poll_interval: timedelta = timedelta(seconds=2)) -> ImportResult:
+        'poll a started import task ID, blocking until completion'
+        ...
+    def run(self, poll_interval: Optional[timedelta] = None) -> ImportResult:
+        'same as calling start() and then wait()'
+        ...
+
+
+class ImportResult(NamedTuple):
+    commit: Commit
+    ingested_objects: int
+
+class Transaction(Branch):
+    def __init__(self, repository_id: str, branch_id: str, client: Client = DefaultClient): ...
+    def begin() -> None:
+        'Create ephemeral branch from the source branch (e.g. <source_branch_id>-txn-<uuid>'
+        ...
+
+    def commit() -> Commit:
+        'commit, merge, delete ephemeral branch'
+        ...
+
+    def rollback(delete_temp_branch: bool = True) -> None:
+        'if delete_temp_branch = True, delete the ephemeral branch createds'
+
+    def __enter__(self):
+        'calls begin()'
+        ...
+
+    def __exit__(self, type, value, traceback):
+        'if successful, commit(), otherwise rollback() and report a meaningful error'
+        ...
+
+```
+
+While this list is fairly exhaustive, it might require a few additional tweaks and additions.
+
+Additionally, we define the following exception hierarchy:
+
+```python
+
+# lakefs.exceptions
+class LakeFSException(Exception):
+    status_code: int
+    message: str
+
+class NotFoundException(LakeFSException): ...
```

Comment:
This would ideally _also_ be a `FileNotFoundException`, but I'm not sure that the Python type system is rich enough to handle this.

Reply:
Added an `ObjectNotFoundException` that inherits both `NotFoundException` and python's `FileNotFoundError`.